sklearn/utils/extmath.py

```diff
@@ -148,7 +149,10 @@ def safe_sparse_dot(a, b, dense_output=False):
         else:
             ret = np.dot(a, b)
     else:
-        ret = a @ b
+        if isinstance(b, LinearOperator) and not isinstance(a, LinearOperator):
+            ret = (b.T @ a.T).T
```

Comment:
Can you explain this line? `a @ b` does not work? Do we actually need to handle this case? I'm OK with the rest of changes, but if we can avoid this special case it would be better IMO.

Reply:
> Can you explain this line? `a @ b` does not work? Do we actually need to handle this case? I'm OK with the rest of changes, but if we can avoid this special case it would be better IMO.

Yes, `a @ b`  doesn't work if `b` is a `LinearOperator` and `a`  is not. Because the first argument of the `@` operator defines actual function that do the job.  `LinearOperator` can be multiplied by anything, but `ndarray` or sparse matrices  know nothing about `LinearOperator` and do not know how to proceed.

Reply:
And so this is necessary for  `randomized_svd`?

Reply:
> And so this is necessary for `randomized_svd`?

Definitely. LinearOperator can incapsulate anything that behave like a matrix via matvec() implementation. Obvious examples are SVD decomposition, einsum() routine, any tensor network with two free legs.

Reply:
Why not return `b@a` then, it should work since here `b` and `a` are 1d vectors?

Reply:
> Why not return `b@a` then, it should work since here `b` and `a` are 1d vectors?

They can be 2D.